RPED can now install certain stack components and display more verbose part information

SyncIt21 opened the pull request:
**Note :**  Rehashed & cleaned up version of PR [#71064](https://github.com/tgstation/tgstation/pull/71064). That one contained unrelated commits and was so bad to the point it had be labeled with DNM. This is the clean version of it. 

## About The Pull Request
Rapid part exchange device RPED & Bluespace version of it BRPED can now

A) pick up 4 types of materials
  - glass sheets[max 30]
  - plasteel[max 30]
  - cable coil[max 30]
  - Bluespace[ore,refined,artificial,stack] total sum of all these should not exceed 30
 and install them in an incomplete machine frame

B)display more verbose part information[number of parts for each type including for stack components]

## Why It's Good For The Game

A)Some machines types such as

- [thermo machines](https://tgstation13.org/wiki/Machines#Freezer.2FHeater)
- [electrolyzers](https://tgstation13.org/wiki/Machines#Electrolyzer)
- crystallizer=10 cable pieces , 10 glass sheets , 5 plasteel sheets

All require either cable ,glass or plasteel. Usually you have to carry these components inside your backpack which occupy space there and later take it out from there and install it manually by hand and then use RPED or BRPED on the machine frame.

This feature now combines these 2 steps into 1 step thus making installing machine frames much faster and you also saves space inside your backpack as these are now carried inside the RPED itself

You need to insert the exact stack type capacity available inside the RPED for it to work i.e. if you pick up 5 glass sheets then later you need a stack of exactly 25 glass sheets to complete the stack. This is easier from a coding perspective so we don't have to deal with subtracting & updating appearances of the remaining stack on the ground or the user hand. 

B) When using the RPED or BRPED to display machine part information these are the new formats

Using an RPED on an thermo machine
![Screenshot (42)](https://user-images.githubusercontent.com/110812394/199978302-bf2e16a8-4f73-4467-be60-4a4fa2ef6468.png)

Using an RPED on the crystallizer
![Screenshot (43)](https://user-images.githubusercontent.com/110812394/199978628-6fb0bf2b-978c-490d-a2e1-e5d1d72b460d.png)

##Change Log
:cl:
add: checks while inserting stack materials into an RPED and their amounts 
code: sorting of RPED part components to sort stack materials by their ratings 
del: old way of displaying parts when clicking on an machine frame with it
add:new way of displaying part info which displays  number of parts including for stack materials 
/:cl:

SyncIt21 committed:
RPED & BRPED picks up stacks & more informative

1)They can now pick up cable ,plasteel & glass
2)Displays more verbose part information such as part numbers

SyncIt21 committed:
Checks for stack types

Checks for the type of stacks the user is trying to insert into the RPED

SyncIt21 committed:
Created wrong directory . Now deleted

Wrong directory _DEFINES created by mistake please ignore

SyncIt21 committed:
Merge branch 'tgstation:master' into RPED-StackPickUp-PartDisplay

SyncIt21 committed:
Merge branch 'tgstation:master' into RPED-StackPickUp-PartDisplay

SyncIt21 committed:
Now pick up bluespace types

can now pick bluespace crystal types and limit all stack types to max 30

SyncIt21 commented:
Can any maintainer remove the stale label from this?. This PR is finished and i have no pending changes to make

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +9,77 @@
 	max_specific_storage = WEIGHT_CLASS_NORMAL
 	numerical_stacking = TRUE
 
+	var/list/allowed_mat_types=list(
```
Document, also make static

```suggestion
	/// your documentation here
	var/static/list/allowed_mat_types = list(
```

Mothblocks replied:
Also full name--mat -> material

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +9,77 @@
+		/obj/item/stack/cable_coil
+	)
+
+	var/list/allowed_bluespace_types=list(
```
```suggestion
	var/static/list/allowed_bluespace_types = list(
```

Document this too

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +9,77 @@
+	var/list/allowed_mat_types=list(
+		/obj/item/stack/sheet/glass,
+		/obj/item/stack/sheet/plasteel,
+		/obj/item/stack/cable_coil
```
```suggestion
		/obj/item/stack/cable_coil,
```

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +9,77 @@
+		/obj/item/stack/ore/bluespace_crystal,
+		/obj/item/stack/ore/bluespace_crystal/refined,
+		/obj/item/stack/ore/bluespace_crystal/artificial,
+		/obj/item/stack/sheet/bluespace_crystal
```
```suggestion
		/obj/item/stack/sheet/bluespace_crystal,
```

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +9,77 @@
+		if(!is_type_in_list(to_insert,allowed_mat_types) && !is_type_in_list(to_insert,allowed_bluespace_types))
+			return FALSE
+
+		var/obj/item/stack/the_stack=to_insert
```
Put a space in between equals signs for all of these

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +9,77 @@
+		//how much of this stack type is currently in the stack
+		var/present_amount=0
+		//what is the maximum allowed amount for this stack type in storage
+		var/max_amount=30
```
Did you mean to have these declared on the typepath? Right now these only exist here and cannot be modified.

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +9,77 @@
 	. = ..()
-	if(!to_insert.get_part_rating())
+
+	///we check how much of glass,plasteel & cable the user can insert
```
```suggestion
	//we check how much of glass,plasteel & cable the user can insert
```

Three slashes = doc comment

SyncIt21 replied:
Thanks. didn't know about that

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +9,77 @@
+
+	///we check how much of glass,plasteel & cable the user can insert
+	if(isstack(to_insert))
+		///user tried to insert invalid stacktype
```
```suggestion
		//user tried to insert invalid stacktype
```

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +9,77 @@
+			to_chat(usr,span_alert("You can only insert exact [available] more of [to_insert.name]!"))
+			return FALSE
+
+
```
Don't have all this empty space

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +9,77 @@
+
+		///no more storage for this specific stack type
+		if(max_amount-present_amount==0)
+			to_chat(usr,span_alert("No more [to_insert.name] can be added!"))
```
Does any other `can_insert` send to_chats?

Mothblocks replied:
Also use `user`, not `usr`.

SyncIt21 replied:
Only this one display chat, your right that does seem off a bit. Ill remove the chat. Renamed usr->user

Mothblocks commented on code/game/machinery/_machinery.dm:
```diff
@@ -955,11 +955,38 @@
 /obj/machinery/proc/display_parts(mob/user)
-	. = list()
-	. += span_notice("It contains the following parts:")
+	var/init_value=0
```
```suggestion
	var/init_value = 0
```

It's not obvious what this refers to by the way

SyncIt21 replied:
It's just a value used to initialize the starting count for a particular part/stack. I put a comment now specifying that

Mothblocks commented on code/game/machinery/_machinery.dm:
```diff
@@ -955,11 +955,38 @@
+			part_count[C.name]++
+
+	var/count=0
+	var/suffix=""
```
Spaces with equals

Mothblocks commented on code/game/machinery/_machinery.dm:
```diff
@@ -955,11 +955,38 @@
+	var/list/printed_components = list()
+
+	var/text = span_notice("It contains the following parts:")
+	for(var/obj/item/C in component_parts)
```
Full name--`component_part` or similar.

Mothblocks commented on code/game/machinery/_machinery.dm:
```diff
@@ -955,11 +955,38 @@
+	var/text = span_notice("It contains the following parts:")
+	for(var/obj/item/C in component_parts)
+		if(!printed_components[C.name])
+			count=part_count[C.name]
+			if(count>1)
+				if(!istype(C,/obj/item/stack/cable_coil) && isstack(C))
+					suffix=" sheets"
+				else
+					suffix="s"
+			else
+				suffix=""
+			text+= span_notice("[icon2html(C, user)] [count] [C.name][suffix].")
```
```suggestion
			text += span_notice("[icon2html(C, user)] [count] [C.singular_name]\s.")
```

Mothblocks commented on code/modules/research/stock_parts.dm:
```diff
@@ -217,7 +217,10 @@ If you create T5+ please take a pass at mech_fabricator.dm. The parts being good
 /proc/cmp_rped_sort(obj/item/A, obj/item/B)
-	return B.get_part_rating() - A.get_part_rating()
+	///all stacks have a rating of 1
+	var/A_Rating=isstack(A)?1:A.get_part_rating()
+	var/B_Rating=isstack(B)?1:B.get_part_rating()
```
Spacing

Also `a_rating`, `b_rating`

Mothblocks commented on code/modules/research/stock_parts.dm:
```diff
@@ -217,7 +217,10 @@ If you create T5+ please take a pass at mech_fabricator.dm. The parts being good
 
 /proc/cmp_rped_sort(obj/item/A, obj/item/B)
-	return B.get_part_rating() - A.get_part_rating()
+	///all stacks have a rating of 1
```
This comment shouldn't exist as it does not explain WHY it's like this

SyncIt21 committed:
@Mothblocks all changes implemented

1)Renamed C to component_part
2)Put space between = sign
3)Made list variables static and added documentation to them
4)Added comma after last item in list

SyncIt21 committed:
@Mothblocks : Cleaned up comments

1) proper multiline comments
2) removed extra slash from all comments

SyncIt21 committed:
@Mothblocks : removed newlines

Removed extra new lines

SyncIt21 committed:
Merge branch 'master' into RPED-StackPickUp-PartDisplay

SyncIt21 committed:
@MothBlocks : Variables fixes

1) Made max amount into a global define
2)More space between = sign

SyncIt21 committed:
Merge branch 'RPED-StackPickUp-PartDisplay' of https://github.com/SyncIt21/The-TG-Station-Fork into RPED-StackPickUp-PartDisplay

SyncIt21 committed:
@MothBlocks : comments & variables

1. added a comment for init_value
2. Even more spacing in the = sign

SyncIt21 commented:
@Mothblocks i think that's all of them. Anything else?

SyncIt21 committed:
Merge branch 'master' into RPED-StackPickUp-PartDisplay

Mothblocks commented:
In the future please don't put pings like that in commits, they're somewhat annoying to mute 😅

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +13,81 @@
+			return FALSE
+
+		var/obj/item/stack/the_stack = to_insert
+		//how much of the stack is the user trying to insert
```
You don't need to comment everything, only what's not obvious

SyncIt21 replied:
Yup i love details my bad ill remove that

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +13,81 @@
+		//stacks type
+		var/obj/item/stack/things
+
+		//not a real location so dont bother
```
This comment isn't helpful because the code makes this obvious, comments should explain why, not what

SyncIt21 replied:
Yup

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +13,81 @@
+	else if(!to_insert.get_part_rating())
 		return FALSE
+
+	return TRUE
```
```suggestion
	return .
```

The parent `can_insert()` might be doing some other logic

SyncIt21 replied:
That return looks silly though. But anyway sure

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +13,81 @@
+
+	/**
+	 * we check if the user is trying to insert any of these bluespace crystal types into the RPED
+	 * at any point the total sum of all these types in the RPED must be 30
```
Where does the number 30 come from? You should reference the variable instead of a flat number, as that might change

SyncIt21 replied:
It comes from the MAX_PICKUP_TYPE variable.  I'll update the comment

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +13,81 @@
+				if(is_type_in_list(things,allowed_bluespace_types))
+					present_amount+=things.amount
+			else if(istype(to_insert,things.type))
+				present_amount=things.amount
```
Why does this override the amount instead of adding to it?

SyncIt21 replied:
Good question.  There are 4 types of Bluespace crystals(see the available_bluespace_types list variable). Let's say the RPED contains 15 Bluespace sheets and 10 refined Bluespace crystals.  Then if the user is trying to pick up raw Bluespace ore we want the user to be able to pickup only 5 ore so the total amount of Bluespace stuff in the rped is 30 I.e(15 sheets+10 refined+5 ore) hence I calculate the total sum of all Bluespace types in the rped and use that as the present amount.  I do this so the user doesn't carry 30 of each Bluespace type in the rped which could only mean bad stuff

Mothblocks commented on code/game/machinery/_machinery.dm:
```diff
@@ -955,11 +955,39 @@
-		. += span_notice("[icon2html(C, user)] \A [C].")
-	. = jointext(., "")
+	//the initial value for an particular part/stack type. We count up from here when more instances are found
+	var/init_value = 0
```
You shouldn't declare this all the way outside of the loop if it's only used inside the loop.

Better yet, you don't need a variable at all, directly set the index instead.

SyncIt21 replied:
Good point. I'll change that

Mothblocks commented on code/game/machinery/_machinery.dm:
```diff
@@ -955,11 +955,39 @@
+					suffix = "s"
+			else
+				suffix=""
+			text+= span_notice("[icon2html(component_part, user)] [count] [component_part.name][suffix].")
```
You didn't apply my change here

SyncIt21 replied:
oooffff

Mothblocks replied:
You still didn't apply my change here :(

SyncIt21 replied:
Small question.  If I remove the check for cable coil then it would print 10 glasses,  10 plasteels and not 10 glass sheets, 10 plasteel sheets. 
That's looks a little silly is that okay?

SyncIt21 replied:
Anyway as u wish

SyncIt21 replied:
![Screenshot (49)](https://user-images.githubusercontent.com/110812394/203708005-1e6d8f22-179e-4a44-a681-de62d2f3ea87.png)

This looks wack but alright

SyncIt21 replied:
Or better yet i'll remove the suffix altogether that will fix it

SyncIt21 replied:
![Screenshot (50)](https://user-images.githubusercontent.com/110812394/203709212-ed4f8b76-0a9b-42b7-8f5e-397aed7b32b1.png)

Alright this looks ok too. I'll just go with this

Mothblocks commented on code/modules/research/stock_parts.dm:
```diff
@@ -217,7 +217,13 @@ If you create T5+ please take a pass at mech_fabricator.dm. The parts being good
+	 * so we assign them an default rating of 1 when the RPED sorts these components
+	 */
+	var/a_rating=isstack(A)?1:A.get_part_rating()
+	var/b_rating=isstack(B)?1:B.get_part_rating()
```
Same reviews still apply here

SyncIt21 replied:
Big ooooffff

SyncIt21 commented:
> In the future please don't put pings like that in commits, they're somewhat annoying to mute 😅

Sorry .just wanted to get ur attention that's all

SyncIt21 committed:
Update code/datums/storage/subtypes/rped.dm

Moved list() one line below

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
Update code/datums/storage/subtypes/rped.dm

Removed space after #

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
Update code/datums/storage/subtypes/rped.dm

Moved list() one line after

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +13,81 @@
+		//how much space is available
+		var/available = 0
+		//stacks type
+		var/obj/item/stack/things
```
So in the for loop I can directly do thing.amount?

SyncIt21 replied:
Ah your right just say your next suggestion

SyncIt21 committed:
Corrections

1. Even MORE spacing in the = sign
2. removed init_value
3. modified for loop to stack_component
4. changed some comments to improve understanding hopefully

SyncIt21 committed:
Even more fixes

1. removed some obvious comments
2. made insert_amount into inline

SyncIt21 committed:
Merge branch 'master' into RPED-StackPickUp-PartDisplay

SyncIt21 commented:
@Mothblocks anything i missed?

SyncIt21 committed:
Merge branch 'master' into RPED-StackPickUp-PartDisplay

SyncIt21 committed:
Merge branch 'master' into RPED-StackPickUp-PartDisplay

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +13,70 @@
+		var/obj/item/stack/the_stack = to_insert
+		var/present_amount = 0
+
+		//not a real location so dont bother
```
```suggestion
```

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +13,70 @@
+				break
+
+		//no more storage for this specific stack type
+		if(MAX_STACK_PICKUP-present_amount==0)
```
```suggestion
		if(MAX_STACK_PICKUP - present_amount == 0)
```

Mothblocks commented on code/datums/storage/subtypes/rped.dm:
```diff
@@ -9,7 +13,70 @@
+	//we check how much of glass,plasteel & cable the user can insert
+	if(isstack(to_insert))
+		//user tried to insert invalid stacktype
+		if(!is_type_in_list(to_insert,allowed_material_types) && !is_type_in_list(to_insert,allowed_bluespace_types))
```
```suggestion
		if(!is_type_in_list(to_insert, allowed_material_types) && !is_type_in_list(to_insert, allowed_bluespace_types))
```

Mothblocks commented on code/game/machinery/_machinery.dm:
```diff
@@ -955,11 +955,36 @@
+
+	var/text = span_notice("It contains the following parts:")
+	for(var/obj/item/component_part in component_parts)
+		if(!printed_components[component_part.name])
```
Early-return

SyncIt21 replied:
Printed components is used just an temporary list to check which components were already printed for example if I have 2 matter bins then I set printed components["matter bin"]=true so the 2nd matter bin in the next iteration  won't get printed again.
If I return here then It would not print remaining components in the foor loop and would break the code

Mothblocks replied:
Early-continue then

SyncIt21 replied:
Ummm that's would break the counting  logic completely  I'll give u an example why. 
Say I have 2 matter bins if I applied your logic and printed_components is empty 
if(!printed_components["matter_bin"])
  continue 
Then the 1st Matter bin is skipped and not printed beacause the list is empty and initially does not have this entry 
Same thing for the 2nd matter bin it's skipped and not printed so that would be a logical error

Mothblocks replied:
You would invert it

```dm
if (printed_components[name])
    part_count[name]++
    continue
```

SyncIt21 replied:
That's correct however part_count[name]++ isn't necessary because it was already initialzed in the previous for loop and should only be printed now so it would be 

if (printed_components[name])
    continue  //already printed dint print again 

//print stuff 

I'll do that

Mothblocks commented on code/game/machinery/_machinery.dm:
```diff
@@ -955,11 +955,36 @@
-	. = jointext(., "")
+	var/list/part_count = list()
+	for(var/obj/item/component_part in component_parts)
+		if(!part_count[component_part.name])
```
Early-return

SyncIt21 replied:
part_count is an temporary map where the key is the part name and the value is its amount in the machine e.g part_count["matter bin"]=2 means there are 2 matter bins. 

The logic is "If this is my first time encountering this part(I.e !part_count) them set its amount to 1 else if have encountered this part before then increment that count

Again same as my previous comment if I return here then It would not count remaining parts in the for loop and the code would break.  

The for loop must complete for all parts to be counted and can't be broken anywhere

Mothblocks replied:
Early-continue is what I meant, which is possible

Mothblocks reviewed (changes requested):
These + my first change with the component printing to use singular_name and not check for cable coil

SyncIt21 committed:
Update code/datums/storage/subtypes/rped.dm

Spacing

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
Update code/datums/storage/subtypes/rped.dm

Spacing

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
Update code/datums/storage/subtypes/rped.dm

Even more spacing

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
Update code/datums/storage/subtypes/rped.dm

Removed new line

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
Update code/datums/storage/subtypes/rped.dm

Removed unessassary  comment

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
Update code/modules/research/stock_parts.dm

Even more spacing

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
Merge branch 'master' into RPED-StackPickUp-PartDisplay

SyncIt21 committed:
Even more more more more......

1. Removed the suffix & cable coil check
2. Inverted if condition

SyncIt21 committed:
Merge branch 'master' into RPED-StackPickUp-PartDisplay

SyncIt21 commented:
> These + my first change with the component printing to use singular_name and not check for cable coil

@Mothblocks now??

SyncIt21 committed:
Merge branch 'master' into RPED-StackPickUp-PartDisplay

Mothblocks commented on code/game/machinery/_machinery.dm:
```diff
@@ -955,11 +955,27 @@
+	for(var/obj/item/component_part in component_parts)
+		if(printed_components[component_part.name])
+			continue //already printed so skip
+		text += span_notice("[icon2html(component_part, user)] [part_count[component_part.name]] [component_part.name].")
```
```suggestion
		text += span_notice("[icon2html(component_part, user)] [part_count[component_part.name]] [component_part.singular_name]\s.")
```

Mothblocks replied:
```suggestion
		
		var/part_name = component_part.name
		if (isstack(component_part))
			var/obj/item/stack/stack_part = component_part
			part_name = stack_part.singular_name
		
		text += span_notice("[icon2html(component_part, user)] [part_count[component_part.name]] [part_name]\s.")
```

Mothblocks commented on code/game/machinery/_machinery.dm:
```diff
@@ -955,11 +955,27 @@
-	. = jointext(., "")
+	var/list/part_count = list()
+	for(var/obj/item/component_part in component_parts)
+		if(!part_count[component_part.name])
+			if(isstack(component_part))
+				var/obj/item/stack/stack_part = component_part
+				part_count[component_part.name] = stack_part.amount
+			else
+				part_count[component_part.name] = 1
+		else
+			part_count[component_part.name]++
```
```suggestion
		if(part_count[component_part.name])
			part_count[component_part.name]++
			continue
		if(isstack(component_part))
			var/obj/item/stack/stack_part = component_part
			part_count[component_part.name] = stack_part.amount
		else
			part_count[component_part.name] = 1
```

Mothblocks commented on code/game/machinery/_machinery.dm:
```diff
@@ -955,11 +955,27 @@
+	var/list/printed_components = list()
+
+	var/text = span_notice("It contains the following parts:")
+	for(var/obj/item/component_part in component_parts)
```
```suggestion
	for(var/obj/item/component_part as anything in component_parts)
```

Mothblocks committed:
Apply suggestions from code review
